### PR TITLE
Establish a unified data handling model

### LINKED ISSUE #6835: Vectorized Arrow Heads
I posted this in the Bokeh Gmail Group before and Bryan advised to open an issue. So here it is (better late then never):

See my sample code below. I'm drawing a whole bunch of arrows in a plot, all arrows have a line weight (lw), an alpha (a) and an arrow line weight (lw_arrow) -basically the arrow size- and start and end coordinates.

Unfortunately when I feed this information to my plot using source it seems impossible to send the 'a' and 'lw_arrow' into the VeeHead function, and it seems only possible to hard code the variables in VeeHead, as I have done now.

What I would like to do instead is the following: end=VeeHead(fill_color="black", size='lw_arrow', fill_alpha='a', line_alpha='a') where 'lw_arrow' and 'a' are coming from source. 

```
from bokeh.plotting import figure
from bokeh.models import Arrow, VeeHead
from bokeh.models import ColumnDataSource
import pandas as pd
from bokeh.io import curdoc

df = pd.DataFrame(
    {'x1': [48.61416189, 50.76303746, 47.54218554, 46.91226294],
     'y1': [78.99324704, 80.69915616, 79.41805739, 80.28060694],
     'x2': [70.60720053, 67.54310965, 52.38867755, 34.25213619],
     'y2': [52.90955905, 93.23068347, 37.70820205, 52.30176102],
     'lw': [5, 6, 7, 2],
     'a': [1, 1, 0.5, 0.25],
     'lw_arrow': [15, 18, 21, 6],

    })

p = figure(plot_width=600, plot_height=428,
               x_range=[0,100],
               y_range=[0,100],
               logo=None,
               toolbar_location = 'below')


source = ColumnDataSource(data=dict(
        x1=[],
        y1=[],
        x2=[],
        y2=[],
        lw=[],
        a=[],
        lw_arrow=[],

    ))

arrows = Arrow(end=VeeHead(fill_color="black", size=10, fill_alpha=.5, line_alpha=.5),
                   x_start='x1',
                   y_start='y1',
                   x_end='x2',
                   y_end='y2',
                   source=source,
                   line_width='lw',
                   line_alpha='a')

p.add_layout(arrows)

def update():
    """
    update the new graph with the values from selected at select_for_graph by passing it through the source
    :return:
    """

    #for the lines we need the whole dataframe
    source.data = dict(
        x1=df['x1'],
        y1=df['y1'],
        x2=df['x2'],
        y2=df['y2'],
        lw=df['lw'],
        a=df['a'],
        lw_arrow=df['lw_arrow']
    )

curdoc().add_root(p)
curdoc().add_periodic_callback(update,1000)
```

### LINKED ISSUE #10203: [FEATURE] line_dash vectorizable where appropriate
For "multi" glyphs and some annotation, it is currently only possible to specify a single line dash style. It could be possible to make this property "vectorized" so that e.g. different individual lines in a `multi_line` can have different dash styles. 

The slight trickiness here will be that dash styles are expressed as lists of numbers, so vectorizing dash styles means accepting "lists of lists" which is not the case for any other property AFAIK. Additionally there are some string aliases for some of the common simple patterns, so those will also have to be handled.

### LINKED ISSUE #10666: [BUG] Webgl perf regressions in branch-2.3
See https://github.com/bokeh/bokeh/pull/10655 for details

### LINKED ISSUE #10738: [BUG] MultiLine datetime precision loss
```python
from bokeh.models import ColumnDataSource
from bokeh.plotting import figure, output_file, show

output_file("line.html")

data = {'x': [[1606075294619, 1606075294646, 1606075294670, 1606075294695]],
       'y': [[1, 2, 3, 6]]}

source = ColumnDataSource(data=data)

fig = figure(height=250, x_axis_type='datetime')
fig.multi_line(xs='x', ys='y', source=source)

show(fig)
```

Screen Shot 2020-11-28 at 12 51 15 PM

### LINKED ISSUE #10801: [BUG] precision loss in patch update when using timeseries data with multi_line glyph (introduced in v2.2.0)
The patch update functionality became very slow (possibly also incorrect) when working with a `datetime` X axis (i.e. timeseries data) and a `multi_line` glyph starting at version 2.2.0. I am not sure about other glyphs, as I have not tested them.

#### Software version details
**Bokeh:** Bug appeared going from 2.1.1 -> 2.2.0 (I am currently using 2.2.3)
**Python:** 3.8
**OS:** Mac OSX 10.15.7
**Browser:** Chrome 87.0.4280.88

#### Description of expected behavior and the observed behavior
Starting at version 2.2.0 of bokeh, the [patch](https://github.com/bokeh/bokeh/blob/branch-2.3/bokeh/models/sources.py#L530) update method became very slow when working with the `multi_line` glyph. I have created a minimal example based on the [patch_app.py](https://github.com/bokeh/bokeh/blob/main/examples/howto/patch_app.py) that demonstrates this behavior.

In the example, you will see two graphs, each randomly updating via patch. One graph has an integer X-axis, and one has a DateTime X-axis. If you run the code once with bokeh version 2.1.1, you will see that both graphs update at the same rate, but starting on bokeh version 2.2.0, the graph with the integer X-axis will update at the same rate as before, but the graph with the DateTime X-axis will update significantly slower, and also at seemingly random intervals, to a point where it is not very usable.

You will also notice that no random points are updated in the datetime X-axis graph. Instead, the updated points "snap" to other points, almost as if they were grouped-by clusters. This is also something that I have noticed in my real-world application, where the intended points are no longer updated correctly. Below, you will find the example code, as well as some pictures showing the results.

#### Complete, minimal, self-contained example code that reproduces the issue

```python
# This is a Bokeh server app. To function, it must be run using the
# Bokeh server ath the command line:
#
#     bokeh serve --show patch_app.py
#
# Running "python patch_app.py" will NOT work.
import numpy as np

from bokeh.io import curdoc
from bokeh.layouts import gridplot
from bokeh.models import ColumnDataSource
from bokeh.plotting import figure
from datetime import datetime, timedelta


def gen_timeseries(size):
    ui_refresh_rate_ms = timedelta(milliseconds=2000)

    start_timestamp = datetime.now()

    timestamp_counter = start_timestamp + ui_refresh_rate_ms
    datetime_axis = list()
    for i in range(size):
        datetime_axis.append(timestamp_counter)
        timestamp_counter += ui_refresh_rate_ms

    return datetime_axis

# CDS with 1d array elements and a numerical X axis
x = np.linspace(0, 10, 200)
y0 = np.sin(x)
y1 = np.cos(x)
source_numerical = ColumnDataSource(data=dict(xs=[x, x], ys=[y0, y1], color=["olive", "navy"]))

p1 = figure(plot_width=400, plot_height=400)
p1.multi_line('xs', 'ys', alpha=0.6, line_width=4, color="color", source=source_numerical)

# CDS with 1d array elements and a datetime X axis
x = gen_timeseries(200)
x_random = np.random.uniform(10, size=200)
y0 = np.sin(x_random)
y1 = np.cos(x_random)
source_datetime = ColumnDataSource(data=dict(xs=[x, x], ys=[y0, y1], color=["red", "green"]))

p2 = figure(plot_width=400, plot_height=400)
p2.multi_line('xs', 'ys', alpha=0.6, line_width=4, color="color", source=source_datetime)


def update():
    # update a single point of the numerical X-axis data
    i = np.random.randint(200)
    new_y_a = 100 * np.random.random()
    new_y_b = 100 * np.random.random()
    source_numerical.patch({'ys': [([0, i], [new_y_a]), ([1, i], [new_y_b])]})

    # update a single point of the datetime X-axis data
    i = np.random.randint(200)
    new_y_a = 100 * np.random.random()
    new_y_b = 100 * np.random.random()
    source_datetime.patch({'ys': [([0, i], [new_y_a]), ([1, i], [new_y_b])]})


curdoc().add_periodic_callback(update, 50)

curdoc().add_root(gridplot([[p1, p2]]))
```

#### Example Screenshots

**bokeh-2.1.1 after 10sec**:
image

**bokeh-2.2.3 after 10sec**:
image

You can see how the identical code results in two very different update patterns for the datetime X-axis graph (right graph) in the two bokeh versions.

I have looked a little bit at the bokeh code to see what the issue could be, but I could not really find anything.

### PR DESCRIPTION
**Pull request by mattpap**
- [x] fixes #10666 
- [x] fixes #10738 
- [x] fixes #10801 
- [x] fixes #10203
- [x] fixes #6835

### CODE REVIEW

File: bokehjs/src/lib/models/grids/grid.ts [hunk 1]
```diff
@@ -40,12 +38,18 @@ export class GridView extends GuideRendererView {
...
+      ctx.rect(sx0[0], sy0[0], sx1[1] - sx0[0], sy1[1] - sy0[0])
+
+      if (this.visuals.band_fill.doit) {
+        this.visuals.band_fill.set_value(ctx)
+        ctx.fill()
+      }
 
-      this.visuals.band_hatch.doit2(ctx, () => {
-        ctx.fillRect(sx0[0], sy0[0], sx1[1] - sx0[0], sy1[1] - sy0[0])
-      }, () => this.request_render())
```
Comment (bryevdv) [inline bokehjs/src/lib/models/grids/grid.ts:48]: What causes this to no longer be necessary? It was there because hatch fills can be remote images that load asynchronously and previously they would not show up without explicitly scheduling a new render

Comment (mattpap) [reply bokehjs/src/lib/models/grids/grid.ts:48]: This will be handled the same way as images are loaded in `ImageURL` glyph, unifying all types of remote resource loading and it will fix readiness notification protocol, which this didn't implement.

Comment (bryevdv) [reply bokehjs/src/lib/models/grids/grid.ts:48]: just making sure I understand, this is still TBD? I looked for something like what `ImageURL` does and did not see it

**Comment by bryevdv** [comment]
@mattpap how far along is this? How can I help? I have several folks getting antsy with me about a 2.3 release for fixes they urgently need. I think this PR needs to be the top priority and then we release as soon as it is ready and push anything else left  off.

**Comment by mattpap** [comment]
I should get this finished by tomorrow.

File: bokeh/models/annotations.py [hunk 2]
```diff
@@ -133,6 +136,16 @@ class Annotation(Renderer):
 
     level = Override(default="annotation")
 
+@abstract
+class DataAnnotation(Annotation):
+    ''' Base class for data annotation models.
```
Comment (bryevdv) [inline bokeh/models/annotations.py:141]: We could be explicit here and explain that a `DataAnnotation` means that it has a `source` property

**Comment by mattpap** [comment]
This is tentatively ready. There are some rough edges and some code duplication, which I will address in future, but for what this PR does, the main bundle increased in size only by 14 kB. This PR is functionally almost what I intended. What's missing is support for inheritance of all non-coordinate data (e.g. radius) by derived glyphs. I have this implemented and it mostly works, but there are quite a few corner cases still to consider, so I don't what to stall this PR unnecessarily. I will submit that work separately. Tests are needed, especially for new feature (e.g. dash vectorization), but that can be done in concurrent PRs. I verified that examples render (including extensions) and that server works, though I didn't test everything. Performance seems to be restored to pre-2.2 level and I verified all code paths, so that no more precision loss should occur (the approach is pretty conservative, which I will refine in future work).

**Comment by mattpap** [comment]
I think that a walk-through meeting may be useful to speed up review of this PR. It's a large PR, but that's mostly because all glyphs, annotations and other models using visuals had to be changed.

**Comment by bryevdv** [comment]
I'll spend quite a bit of time with this today. Feel free to send a calendar invite, I'm mostly free all week.

**Comment by bryevdv** [comment]
@mattpap The `webgl/marker_compare.py` generates colors like this:
```
colors = np.random.randint(0, 255, size=(N, 3), dtype="uint8")

source = ColumnDataSource(data=dict(x=x, y=y, colors=colors))
```
But the result is coming out grayscale:

Screen Shot 2021-01-18 at 1 16 11 PM

**Comment by bryevdv** [comment]
@mattpap not sure it's related to this PR but the layout on the `dash` example is broken pretty badly, with the rangetool plopt on top of the middle of the plot above:

Screen Shot 2021-01-18 at 1 28 24 PM

**Comment by bryevdv** [comment]
@mattpap I confirmed these layout regressions are already present on `branch-2.3` I will make a separate issue

**Comment by bryevdv** [comment]
Very minor weird change on this PR for `Continuous Updating.ipynb` -- used to be if you stopped kernel execution and ran the last cell after the loop, the hover style changed immediately on the paused plot. Now the change does not show up until the loop cell is re-executed.

**Comment by bryevdv** [comment]
OK @mattpap I tested a pretty good cross section of things manually. With the exception noted above, all the regressions I found existed prior to this PR so I made separate issues. I think it would be good to merge this ASAP and get a new dev build out to make it easier for others to test.

**Comment by mattpap** [comment]
> I confirmed these layout regressions are already present on branch-2.3 I will make a separate issue

That's definitively fallout from color bar layout changes.

**Comment by mattpap** [comment]
I resolved merge conflicts.

**Review by bryevdv — APPROVED** [review]

**Comment by mattpap** [comment]
> Very minor weird change on this PR for Continuous Updating.ipynb -- used to be if you stopped kernel execution and ran the last cell after the loop, the hover style changed immediately on the paused plot. Now the change does not show up until the loop cell is re-executed.

I'm not sure how that worked before, but only main glyph had signals connected in glyph renderer. Anyway, it's a simple fix.
